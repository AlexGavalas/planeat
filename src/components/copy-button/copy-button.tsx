import {
    ActionIcon,
    type ActionIconVariant,
    CopyButton as MantineCopyButton,
    type MantineSize,
    Tooltip,
    type TooltipProps,
} from '@mantine/core';
import { Check, Copy } from 'iconoir-react';
import { useTranslation } from 'next-i18next';

type CopyButtonProps = {
    value: string;
    variant?: ActionIconVariant;
    size?: MantineSize;
    tooltipPosition?: TooltipProps['position'];
};

export const CopyButton = ({
    value,
    variant = 'light',
    size = 'lg',
    tooltipPosition,
}: CopyButtonProps) => {
    const { t } = useTranslation();

    return (
        <MantineCopyButton timeout={2000} value={value}>
            {({ copied, copy: handleCopy }) => (
                <Tooltip
                    withArrow
                    label={copied ? t('util.copied') : t('util.copy')}
                    position={tooltipPosition}
                >
                    <ActionIcon
                        onClick={handleCopy}
                        size={size}
                        variant={variant}
                    >
                        {copied ? <Check /> : <Copy />}
                    </ActionIcon>
                </Tooltip>
            )}
        </MantineCopyButton>
    );
};
