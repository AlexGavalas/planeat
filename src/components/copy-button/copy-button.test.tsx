import { renderWithUser, screen } from '~test/utils';

import { CopyButton } from './copy-button';

describe('<CopyButton />', () => {
    it('renders the button', () => {
        const { container } = renderWithUser(<CopyButton value="test" />);

        expect(container).toMatchSnapshot();
    });

    describe('when the user hovers the button', () => {
        it('shows the tooltip', async () => {
            expect.hasAssertions();

            const { user, baseElement } = renderWithUser(
                <CopyButton value="test" />,
            );

            const button = screen.getByRole('button');

            await user.hover(button);

            expect(baseElement).toMatchSnapshot();
        });
    });

    describe('when user clicks the button', () => {
        afterEach(() => {
            jest.restoreAllMocks();
        });

        it('copies the value to the clipboard', async () => {
            expect.hasAssertions();

            const writeText = jest.spyOn(navigator.clipboard, 'writeText');

            const { user } = renderWithUser(<CopyButton value="test" />);

            const button = screen.getByRole('button');

            await user.click(button);

            expect(writeText).toHaveBeenCalledWith('test');
        });
    });
});
