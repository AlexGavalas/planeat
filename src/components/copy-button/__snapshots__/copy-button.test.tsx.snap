// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CopyButton /> renders the button 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <button
    class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
    data-size="lg"
    data-variant="light"
    style="--ai-size: var(--ai-size-lg); --ai-bg: var(--mantine-color-blue-light); --ai-hover: var(--mantine-color-blue-light-hover); --ai-color: var(--mantine-color-blue-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
    type="button"
  >
    <svg
      color="currentColor"
      fill="none"
      height="1.5em"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      width="1.5em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19.4 20H9.6a.6.6 0 01-.6-.6V9.6a.6.6 0 01.6-.6h9.8a.6.6 0 01.6.6v9.8a.6.6 0 01-.6.6z"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
      <path
        d="M15 9V4.6a.6.6 0 00-.6-.6H4.6a.6.6 0 00-.6.6v9.8a.6.6 0 00.6.6H9"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </button>
</div>
`;

exports[`<CopyButton /> when the user hovers the button shows the tooltip 1`] = `
<body>
  <div>
    <style
      data-mantine-styles="classes"
    >
      @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
    </style>
    <button
      aria-describedby=":r5:"
      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
      data-size="lg"
      data-variant="light"
      style="--ai-size: var(--ai-size-lg); --ai-bg: var(--mantine-color-blue-light); --ai-hover: var(--mantine-color-blue-light-hover); --ai-color: var(--mantine-color-blue-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <svg
        color="currentColor"
        fill="none"
        height="1.5em"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="1.5em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.4 20H9.6a.6.6 0 01-.6-.6V9.6a.6.6 0 01.6-.6h9.8a.6.6 0 01.6.6v9.8a.6.6 0 01-.6.6z"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
        <path
          d="M15 9V4.6a.6.6 0 00-.6-.6H4.6a.6.6 0 00-.6.6v9.8a.6.6 0 00.6.6H9"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
  <div>
    <div>
      <div
        class="m-1b3c8819 mantine-Tooltip-tooltip"
        id=":r5:"
        role="tooltip"
        style="transition-property: opacity; transition-duration: 100ms; transition-timing-function: ease; opacity: 1; z-index: 300; top: -7px; left: 8px;"
        tabindex="-1"
      >
        Copy
        <div
          class="m-f898399f mantine-Tooltip-arrow"
          style="transform: rotate(45deg); position: absolute; border-bottom-right-radius: 0; border-top-color: transparent; border-left-color: transparent; left: -1px;"
        />
      </div>
    </div>
  </div>
</body>
`;
